File: scripts/deploy-metadata.py
#!/usr/bin/env python3
"""
Bulk-deploy SEO metadata from the external CSV metadata map into
src/data/page-meta.ts.

This script regenerates the entire PAGE_META object from the CSV while
preserving existing ogImage expressions (matched by URL path).
"""

import csv
import re
from pathlib import Path

REPO = Path(__file__).resolve().parent.parent
CSV_PATH = Path("/Users/openclaw/Movies/LIve website/mychef-seo/10-metadata-map.csv")
PAGE_META_PATH = REPO / "src/data/page-meta.ts"

SITE = "https://mychef.id"


def safe_key(url: str) -> str:
    key = re.sub(r"^/+|/+$", "", url)
    key = re.sub(r"[^a-zA-Z0-9]+", "-", key)
    key = re.sub(r"^-|-$", "", key)
    return key or "page"


def key_literal(key: str) -> str:
    if re.match(r"^[A-Za-z_$][\w$]*$", key):
        return key
    return f"'{key.replace(chr(39), '\\' + chr(39))}'"


def escape_single(value: str) -> str:
    return value.replace("\\", "\\\\").replace("'", "\\'")


def sq(value: str) -> str:
    return f"'{escape_single(value)}'"


def parse_existing(path: Path) -> dict:
    text = path.read_text(encoding="utf-8")
    entries = {}

    # Match top-level key followed by { block
    for m in re.finditer(
        r"^\s+([A-Za-z_$][\w$]*|'[^']*'|\"[^\"]*\"):\s*\{", text, re.M
    ):
        key_raw = m.group(1)
        if key_raw.startswith(("'", '"')):
            key = key_raw[1:-1]
        else:
            key = key_raw

        start = text.find("{", m.end() - 1)
        depth = 1
        i = start + 1
        while i < len(text) and depth > 0:
            c = text[i]
            if c in ('"', "'", "`"):
                # skip string/template literal
                quote = c
                i += 1
                while i < len(text):
                    if text[i] == "\\":
                        i += 2
                        continue
                    if text[i] == quote:
                        i += 1
                        break
                    i += 1
                continue
            if c == "{":
                depth += 1
            elif c == "}":
                depth -= 1
                if depth == 0:
                    break
            i += 1

        block = text[start : i + 1]
        path_match = re.search(r"path:\s*(['\"`])(.+?)\1,", block)
        og_match = re.search(r"ogImage:\s*(.+?),\s*\n", block)
        if path_match:
            entries[path_match.group(2)] = {
                "key": key,
                "ogImage": og_match.group(1) if og_match else None,
            }
    return entries


def read_csv(path: Path) -> list[dict]:
    text = path.read_text(encoding="utf-8")
    # csv module handles quoted fields
    reader = csv.DictReader(text.splitlines())
    rows = []
    for row in reader:
        action = (row.get("action") or "").strip().lower()
        if action not in ("rewrite", "keep"):
            continue
        rows.append(row)
    return rows


def canonical_expr(canonical: str, url: str) -> str:
    c = canonical.strip()
    if c == "self":
        return f"`${{SITE}}{url}`"
    if c.startswith("/"):
        return f"`${{SITE}}{c}`"
    return sq(c)


def format_entry(key: str, path: str, title: str, description: str, h1: str, canonical: str, og_image: str | None) -> str:
    lines = [
        f"  {key_literal(key)}: {{",
        f"    path: {sq(path)},",
        f"    title: {sq(title)},",
        f"    description:",
        f"      {sq(description)},",
        f"    canonical: {canonical_expr(canonical, path)},",
        f"    h1: {sq(h1)},",
    ]
    if og_image:
        lines.append(f"    ogImage: {og_image},")
    lines.append("  },")
    return "\n".join(lines)


def main():
    if not CSV_PATH.exists():
        raise FileNotFoundError(f"CSV not found: {CSV_PATH}")

    existing = parse_existing(PAGE_META_PATH)
    rows = read_csv(CSV_PATH)

    # Preserve existing ogImage by path
    og_by_path: dict[str, str | None] = {
        path: data.get("ogImage") for path, data in existing.items()
    }
    existing_key_by_path: dict[str, str] = {
        path: data["key"] for path, data in existing.items()
    }

    csv_paths = set()
    blocks: list[str] = []

    for row in rows:
        url = (row.get("url") or "").strip()
        if not url:
            continue
        csv_paths.add(url)

        title = (row.get("recommended_title") or "").strip()
        description = (row.get("recommended_meta") or "").strip()
        h1 = (row.get("recommended_h1") or "").strip()
        canonical = (row.get("canonical") or "self").strip()

        key = existing_key_by_path.get(url) or safe_key(url)
        og = og_by_path.get(url)

        blocks.append(format_entry(key, url, title, description, h1, canonical, og))

    # Append any existing entries not present in CSV, unchanged
    extra_blocks = []
    for path, data in existing.items():
        if path in csv_paths:
            continue
        # We don't have the full block text, but we can keep the ogImage.
        # Since these are extras, reconstruct with current title/desc/h1 from existing
        # is not possible from parse_existing. Better: preserve raw block.
        # For simplicity, skip reconstructing extras; instead keep them by reading raw.
        pass

    joined = "\n\n".join(blocks)

    # Handle extras by raw preservation
    text = PAGE_META_PATH.read_text(encoding="utf-8")
    extra_sections: list[str] = []
    for m in re.finditer(
        r"^\s+([A-Za-z_$][\w$]*|'[^']*'|\"[^\"]*\"):\s*\{", text, re.M
    ):
        key_raw = m.group(1)
        key = key_raw[1:-1] if key_raw.startswith(("'", '"')) else key_raw
        start = m.start()
        brace_start = text.find("{", m.end() - 1)
        depth = 1
        i = brace_start + 1
        while i < len(text) and depth > 0:
            c = text[i]
            if c in ('"', "'", "`"):
                quote = c
                i += 1
                while i < len(text):
                    if text[i] == "\\":
                        i += 2
                        continue
                    if text[i] == quote:
                        i += 1
                        break
                    i += 1
                continue
            if c == "{":
                depth += 1
            elif c == "}":
                depth -= 1
                if depth == 0:
                    break
            i += 1
        block_end = i + 1
        # include trailing comma
        comma_end = block_end
        while comma_end < len(text) and text[comma_end] in ", \n":
            if text[comma_end] == ",":
                comma_end += 1
                break
            comma_end += 1

        path_match = re.search(r"path:\s*(['\"`])(.+?)\1,", text[brace_start : block_end])
        if path_match:
            p = path_match.group(2)
            if p not in csv_paths:
                extra_sections.append(text[start : comma_end])

    if extra_sections:
        joined += "\n\n" + "\n\n".join(extra_sections)

    new_file = f"""/**
 * myCHEF — PAGE META DATA
 *
 * Centralised, unique title / description / H1 for every primary route.
 * Prevents duplicate title tag catastrophes and keeps SEO data in one
 * typed source of truth.
 */

import {{ LOCATION_LANDING_PAGES }} from './locationLandingPages'

export interface PageMeta {{
  /** Route path (e.g. '/', '/fine-dining') */
  path: string
  /** HTML <title> — must be unique site-wide, ≤ 60 chars ideal */
  title: string
  /** Meta description — ≤ 160 chars ideal */
  description: string
  /** Canonical URL */
  canonical: string
  /** Primary H1 heading — must be unique and keyword-rich */
  h1: string
  /** Open Graph / social image */
  ogImage?: string
}}

const SITE = '{SITE}'

export const PAGE_META: Record<string, PageMeta> = {{
{joined}
}} as const

/** Helper to retrieve meta by route key with strict typing */
export function getPageMeta(key: keyof typeof PAGE_META): PageMeta {{
  return PAGE_META[key]
}}

/** All page meta entries as an array for bulk operations (sitemaps, audits, etc.) */
export const ALL_PAGE_META: PageMeta[] = Object.values(PAGE_META)
"""

    PAGE_META_PATH.write_text(new_file, encoding="utf-8")

    print(f"Metadata deploy complete")
    print(f"  CSV rows deployed: {len(rows)}")
    print(f"  Existing entries preserved (not in CSV): {len(extra_sections)}")


if __name__ == "__main__":
    main()

File: scripts/deploy-metadata.ts
#!/usr/bin/env tsx
/**
 * Bulk-deploy SEO metadata from the external CSV metadata map into
 * src/data/page-meta.ts.
 *
 * For every CSV row whose action is "rewrite" or "keep":
 *  - Update an existing PAGE_META entry matched by URL path or generated key.
 *  - Or append a new entry if the page is not yet present.
 *
 * Existing entries that are not referenced by the CSV are left untouched,
 * including their formatting and any dynamic values (e.g. LOCATION_LANDING_PAGES).
 */

import * as fs from 'fs'
import * as path from 'path'
import { fileURLToPath } from 'url'

const __filename = fileURLToPath(import.meta.url)
const __dirname = path.dirname(__filename)

const CSV_PATH = path.resolve('/Users/openclaw/Movies/LIve website/mychef-seo/10-metadata-map.csv')
const PAGE_META_PATH = path.resolve(__dirname, '../src/data/page-meta.ts')

interface ParsedEntry {
  key: string
  start: number
  end: number
  text: string
  path: string
}

interface MetaUpdate {
  path: string
  title: string
  description: string
  h1: string
  canonicalExpr: string
  ogImageExpr?: string
}

/** Convert a URL path into a safe JS object key ("path/to-page" -> "path-to-page"). */
function safeKeyFromUrl(url: string): string {
  const key = url
    .replace(/^\/+|\/+$/g, '')
    .replace(/[^a-zA-Z0-9]+/g, '-')
    .replace(/^-|-$/g, '')
    .toLowerCase()
  return key || 'page'
}

/** Render a key as an unquoted identifier when possible, otherwise single-quoted. */
function keyLiteral(key: string): string {
  return /^[A-Za-z_$][\w$]*$/.test(key) ? key : `'${escapeSingleQuotes(key)}'`
}

/** Single-quote a string literal, escaping backslashes and single quotes. */
function sq(value: string): string {
  return `'${escapeSingleQuotes(value)}'`
}

function escapeSingleQuotes(value: string): string {
  return value.replace(/\\/g, '\\\\').replace(/'/g, "\\'")
}

/** Simple CSV parser that handles quoted fields and commas inside quotes. */
function parseCSV(text: string): string[][] {
  const rows: string[][] = []
  let row: string[] = []
  let field = ''
  let inQuotes = false

  for (let i = 0; i < text.length; i++) {
    const ch = text[i]

    if (inQuotes) {
      if (ch === '"') {
        if (text[i + 1] === '"') {
          field += '"'
          i++
        } else {
          inQuotes = false
        }
      } else {
        field += ch
      }
      continue
    }

    if (ch === '"') {
      inQuotes = true
    } else if (ch === ',') {
      row.push(field)
      field = ''
    } else if (ch === '\n') {
      row.push(field)
      if (row.length > 1 || row[0] !== '') {
        rows.push(row)
      }
      row = []
      field = ''
    } else if (ch === '\r') {
      // Ignore carriage returns; \n terminates the row.
    } else {
      field += ch
    }
  }

  if (field !== '' || row.length > 0) {
    row.push(field)
    if (row.length > 1 || row[0] !== '') {
      rows.push(row)
    }
  }

  return rows
}

/** Locate every top-level object entry in the PAGE_META body. */
function findEntries(body: string): ParsedEntry[] {
  const entries: ParsedEntry[] = []
  const keyRegex = /([A-Za-z_$][\w$]*|'[^']*'|"[^"]*"):\s*\{/g
  let match: RegExpExecArray | null

  while ((match = keyRegex.exec(body)) !== null) {
    const rawKey = match[1]
    const key =
      rawKey.startsWith("'") || rawKey.startsWith('"')
        ? rawKey.slice(1, -1)
        : rawKey

    const blockStart = body.indexOf('{', match.index)
    const keyStart = match.index
    const end = findClosingBrace(body, blockStart)
    if (end === -1) {
      throw new Error(`Could not find closing brace for PAGE_META key "${key}"`)
    }

    const block = body.slice(blockStart, end + 1)
    const pathMatch = block.match(/path:\s*(['"`])(.*?)\1,/s)

    entries.push({
      key,
      start: keyStart,
      end: end + 1,
      text: block,
      path: pathMatch?.[2] ?? '',
    })
  }

  return entries
}

/** Find the matching `}` for the `{` at `start`, respecting strings and template literals. */
function findClosingBrace(text: string, start: number): number {
  let depth = 1
  let i = start + 1
  let inString: null | '"' | "'" | '`' = null
  let escape = false
  let templateDepth = 0

  while (i < text.length) {
    const c = text[i]

    if (escape) {
      escape = false
      i++
      continue
    }

    if (c === '\\') {
      escape = true
      i++
      continue
    }

    if (inString) {
      if (c === '$' && inString === '`' && text[i + 1] === '{') {
        templateDepth++
        i += 2
        continue
      }

      if (c === '}' && inString === '`' && templateDepth > 0) {
        templateDepth--
        i++
        continue
      }

      if (c === inString && templateDepth === 0) {
        inString = null
      }

      i++
      continue
    }

    if (c === '"' || c === "'" || c === '`') {
      inString = c
      i++
      continue
    }

    if (c === '{') {
      depth++
    } else if (c === '}') {
      depth--
      if (depth === 0) {
        return i
      }
    }

    i++
  }

  return -1
}

function extractOgImageExpr(block: string): string | undefined {
  const match = block.match(/ogImage:\s*(.+?)\s*,/s)
  return match?.[1]
}

function buildCanonicalExpr(canonical: string, url: string): string {
  const c = canonical.trim()
  if (c === 'self') {
    return `\`${SITE_PLACEHOLDER}${url}\``
  }
  if (c.startsWith('/')) {
    return `\`${SITE_PLACEHOLDER}${c}\``
  }
  return sq(c)
}

const SITE_PLACEHOLDER = '${SITE}'

function formatEntry(key: string, meta: MetaUpdate): string {
  const lines: string[] = []
  lines.push(`  ${keyLiteral(key)}: {`)
  lines.push(`    path: ${sq(meta.path)},`)
  lines.push(`    title: ${sq(meta.title)},`)
  lines.push(`    description:`)
  lines.push(`      ${sq(meta.description)},`)
  lines.push(`    canonical: ${meta.canonicalExpr},`)
  lines.push(`    h1: ${sq(meta.h1)},`)
  if (meta.ogImageExpr) {
    lines.push(`    ogImage: ${meta.ogImageExpr},`)
  }
  lines.push(`  },`)
  return lines.join('\n')
}

function main() {
  if (!fs.existsSync(CSV_PATH)) {
    console.error(`CSV not found: ${CSV_PATH}`)
    process.exit(1)
  }

  const csvText = fs.readFileSync(CSV_PATH, 'utf8')
  const csvRows = parseCSV(csvText)

  if (csvRows.length === 0) {
    console.error('CSV appears to be empty')
    process.exit(1)
  }

  const pageMetaText = fs.readFileSync(PAGE_META_PATH, 'utf8')

  const declMatch = pageMetaText.match(/export const PAGE_META: Record<string, PageMeta> = \{/)
  if (!declMatch || declMatch.index === undefined) {
    console.error('Could not find PAGE_META declaration in page-meta.ts')
    process.exit(1)
  }

  const closeMatch = pageMetaText.match(/\n\} as const/)
  if (!closeMatch || closeMatch.index === undefined) {
    console.error('Could not find PAGE_META closing "} as const" in page-meta.ts')
    process.exit(1)
  }

  const bodyStart = declMatch.index + declMatch[0].length
  const bodyEnd = closeMatch.index
  const body = pageMetaText.slice(bodyStart, bodyEnd)

  const entries = findEntries(body)
  const byPath = new Map(entries.map((e) => [e.path, e]))
  const byKey = new Map(entries.map((e) => [e.key, e]))

  const updates = new Map<string, { entry: ParsedEntry; block: string }>()
  const additions: { key: string; block: string }[] = []
  let skipped = 0

  // Skip the header row.
  for (let i = 1; i < csvRows.length; i++) {
    const row = csvRows[i]
    if (row.length < 11) {
      skipped++
      continue
    }

    const [
      url,
      _currentTitle,
      recommendedTitle,
      _currentMeta,
      recommendedMeta,
      recommendedH1,
      canonical,
      _ogTitle,
      _ogDescription,
      _schema,
      action,
    ] = row

    const actionNorm = action.trim().toLowerCase()
    if (actionNorm !== 'rewrite' && actionNorm !== 'keep') {
      skipped++
      continue
    }

    const existing = byPath.get(url) ?? byKey.get(safeKeyFromUrl(url))
    const key = existing?.key ?? safeKeyFromUrl(url)
    const ogImageExpr = existing ? extractOgImageExpr(existing.text) : undefined

    const update: MetaUpdate = {
      path: url,
      title: recommendedTitle,
      description: recommendedMeta,
      h1: recommendedH1,
      canonicalExpr: buildCanonicalExpr(canonical, url),
      ogImageExpr,
    }

    const block = formatEntry(key, update)

    if (existing) {
      updates.set(existing.key, { entry: existing, block })
    } else {
      additions.push({ key, block })
    }
  }

  // Rebuild the PAGE_META body, preserving the order and formatting of
  // untouched entries.
  let pos = 0
  let newBody = ''

  for (const entry of entries) {
    const updated = updates.get(entry.key)
    if (updated) {
      newBody += body.slice(pos, entry.start) + updated.block
      pos = entry.end
    } else {
      pos = entry.end
    }
  }

  newBody += body.slice(pos)

  // Append brand-new entries before the final object-closing `}`.
  if (additions.length > 0) {
    const closeIdx = newBody.lastIndexOf('}')
    if (closeIdx === -1) {
      throw new Error('Could not locate the final closing brace of PAGE_META')
    }

    const prefix = newBody.slice(0, closeIdx)
    const suffix = newBody.slice(closeIdx + 1)
    const inserted = additions
      .map((addition, idx) => {
        return idx === 0 ? `\n\n${addition.block}` : `,\n${addition.block}`
      })
      .join('')

    newBody = `${prefix}${inserted}\n}${suffix}`
  }

  let newFile =
    pageMetaText.slice(0, bodyStart) + newBody + pageMetaText.slice(bodyEnd)

  // The replacement preserves the original trailing comma after each entry;
  // formatEntry also appends one. Collapse any doubled commas that result.
  newFile = newFile.replace(/},\s*,/g, '},')

  fs.writeFileSync(PAGE_META_PATH, newFile, 'utf8')

  console.log('Metadata deploy complete')
  console.log(`  Updated existing entries: ${updates.size}`)
  console.log(`  Added new entries:        ${additions.length}`)
  console.log(`  Skipped rows:             ${skipped}`)
}

main()
